### ISSUE TITLE
Cannot install -r ./requirements.txt (line 30) and PyJWT==2.0.1 because these package versions have conflicting dependencies.

### ISSUE THREAD

**Issue by dvelardez** [DefectDojo/django-DefectDojo#4262]
**Bug description**
Cannot build images locally because these package versions have conflicting dependencies.
The conflict is caused by:
    The user requested PyJWT==2.0.1
    pygithub 1.54.1 depends on pyjwt<2.0

**Steps to reproduce**
docker build --build-arg http_proxy=http://myproxy.com:8080 --build-arg https_proxy=http://myproxy.com:8080 -t defectdojo/defectdojo-django -f Dockerfile.django .

**Expected behavior**
Build successful.

**Deployment method** 
Kubernetes

**Environment information**
 - Operating System: Windows 10
 - DefectDojo Commit Message: [2021-04-08 19:59:44 +0200] 96f2f348: Merge pull request #4234 from DefectDojo/release/1.14.1 [ (HEAD -> master, tag: 1.14.1, origin/master, origin/HEAD)]

**Console logs**

```
[stage-1  8/17] RUN pip3 install     --no-cache-dir  --no-index   --find-links=/tmp/wheels   -r ./requirements.txt:
#17 5.441 Looking in links: /tmp/wheels
#17 5.452 Collecting django-tastypie-swagger
#17 5.453   Cloning https://github.com/DefectDojo/django-tastypie-swagger.git (to revision c7c5629480dc20cadd3015490a01bd8e70393c36) to /tmp/pip-install-wvjvs3ew/django-tastypie-swagger_2315fbf3037b4b86bc91d5bdbc22c360
#17 5.453   Running command git clone -q https://github.com/DefectDojo/django-tastypie-swagger.git /tmp/pip-install-wvjvs3ew/django-tastypie-swagger_2315fbf3037b4b86bc91d5bdbc22c360
#17 8.928   Running command git rev-parse -q --verify 'sha^c7c5629480dc20cadd3015490a01bd8e70393c36'
#17 8.933   Running command git fetch -q https://github.com/DefectDojo/django-tastypie-swagger.git c7c5629480dc20cadd3015490a01bd8e70393c36
#17 9.910 Processing /tmp/wheels/asteval-0.9.23-py3-none-any.whl
#17 9.927 Processing /tmp/wheels/bleach-3.3.0-py2.py3-none-any.whl
#17 9.952 Processing /tmp/wheels/celery-4.4.7-py2.py3-none-any.whl
#17 9.970 Processing /tmp/wheels/coverage-5.5-cp36-cp36m-manylinux2010_x86_64.whl
#17 9.989 Processing /tmp/wheels/defusedxml-0.7.1-py2.py3-none-any.whl
#17 10.01 Processing /tmp/wheels/django_celery_results-2.0.1-py2.py3-none-any.whl
#17 10.03 Processing /tmp/wheels/django_auditlog-1.0a1-py3-none-any.whl
#17 10.05 Processing /tmp/wheels/django_dbbackup-3.3.0-py3-none-any.whl
#17 10.07 Processing /tmp/wheels/django_environ-0.4.5-py2.py3-none-any.whl
#17 10.09 Processing /tmp/wheels/django_filter-2.4.0-py3-none-any.whl
#17 10.10 Processing /tmp/wheels/django_imagekit-4.0.2-py2.py3-none-any.whl
#17 10.12 Processing /tmp/wheels/django_multiselectfield-0.1.12-py3-none-any.whl
#17 10.14 Processing /tmp/wheels/django_polymorphic-3.0.0-py2.py3-none-any.whl
#17 10.16 Processing /tmp/wheels/django_crispy_forms-1.11.2-py3-none-any.whl
#17 10.18 Processing /tmp/wheels/django_extensions-3.1.1-py3-none-any.whl
#17 10.20 Processing /tmp/wheels/django_rest_swagger-2.2.0-py2.py3-none-any.whl
#17 10.22 Processing /tmp/wheels/django_slack-5.16.0-py3-none-any.whl
#17 10.24 Processing /tmp/wheels/django_tagging-0.5.0-py3-none-any.whl
#17 10.26 Processing /tmp/wheels/django_tastypie-0.14.2-py2.py3-none-any.whl
#17 10.28 Processing /tmp/wheels/django_watson-1.5.5-py2.py3-none-any.whl
#17 10.30 Processing /tmp/wheels/django_prometheus-2.1.0-py2.py3-none-any.whl
#17 10.32 Processing /tmp/wheels/Django-2.2.18-py3-none-any.whl
#17 10.36 Processing /tmp/wheels/djangorestframework-3.12.3-py3-none-any.whl
#17 10.38 Processing /tmp/wheels/gunicorn-20.0.4-py2.py3-none-any.whl
#17 10.40 Processing /tmp/wheels/html2text-2020.1.16-py3-none-any.whl
#17 10.42 Processing /tmp/wheels/humanize-3.3.0-py3-none-any.whl
#17 10.44 Processing /tmp/wheels/jira-2.0.0-py2.py3-none-any.whl
#17 10.46 Processing /tmp/wheels/PyGithub-1.54.1-py3-none-any.whl
#17 10.48 Processing /tmp/wheels/lxml-4.6.3-cp36-cp36m-manylinux1_x86_64.whl
#17 10.50 Processing /tmp/wheels/Markdown-3.3.4-py3-none-any.whl
#17 10.52 Processing /tmp/wheels/mysqlclient-2.0.3-cp36-cp36m-linux_x86_64.whl
#17 10.53 Processing /tmp/wheels/openpyxl-3.0.7-py2.py3-none-any.whl
#17 10.55 Processing /tmp/wheels/xlrd-1.2.0-py2.py3-none-any.whl
#17 10.57 Processing /tmp/wheels/pdfkit-0.6.1-py3-none-any.whl
#17 10.59 Processing /tmp/wheels/Pillow-8.1.2-cp36-cp36m-manylinux1_x86_64.whl
#17 10.62 Processing /tmp/wheels/psycopg2_binary-2.8.6-cp36-cp36m-manylinux1_x86_64.whl
#17 10.63 Processing /tmp/wheels/cryptography-3.4.7-cp36-abi3-manylinux2014_x86_64.whl
#17 10.65 Processing /tmp/wheels/python_dateutil-2.8.1-py2.py3-none-any.whl
#17 10.67 Processing /tmp/wheels/pytz-2021.1-py2.py3-none-any.whl
#17 10.70 Processing /tmp/wheels/redis-3.5.3-py2.py3-none-any.whl
#17 10.72 Processing /tmp/wheels/requests-2.25.1-py2.py3-none-any.whl
#17 10.74 Processing /tmp/wheels/SQLAlchemy-1.4.3-cp36-cp36m-manylinux2014_x86_64.whl
#17 10.76 Processing /tmp/wheels/supervisor-4.2.2-py2.py3-none-any.whl
#17 10.78 Processing /tmp/wheels/urllib3-1.26.4-py2.py3-none-any.whl
#17 10.80 Processing /tmp/wheels/uWSGI-2.0.19.1-cp36-cp36m-linux_x86_64.whl
#17 10.82 Processing /tmp/wheels/vobject-0.9.6.1-py3-none-any.whl
#17 10.84 Processing /tmp/wheels/whitenoise-5.2.0-py2.py3-none-any.whl
#17 10.86 Processing /tmp/wheels/titlecase-2.0.0-py3-none-any.whl
#17 10.88 Processing /tmp/wheels/social_auth_app_django-4.0.0-py3-none-any.whl
#17 10.90 Processing /tmp/wheels/social_auth_core-4.1.0-py3-none-any.whl
#17 10.92 Processing /tmp/wheels/python_jose-3.2.0-py2.py3-none-any.whl
#17 10.94 Processing /tmp/wheels/GitPython-3.1.14-py3-none-any.whl
#17 10.96 Processing /tmp/wheels/ptvsd-4.3.2-cp36-cp36m-manylinux1_x86_64.whl
#17 10.98 Processing /tmp/wheels/python_gitlab-2.6.0-py3-none-any.whl
#17 11.00 Processing /tmp/wheels/google_auth-1.28.0-py2.py3-none-any.whl
#17 11.02 Processing /tmp/wheels/google_api_python_client-1.12.8-py2.py3-none-any.whl
#17 11.04 Processing /tmp/wheels/google_auth_oauthlib-0.4.4-py2.py3-none-any.whl
#17 11.06 Processing /tmp/wheels/drf_yasg-1.20.0-py2.py3-none-any.whl
#17 11.08 Processing /tmp/wheels/django_saml2_auth-2.2.1-py2.py3-none-any.whl
#17 11.10 Processing /tmp/wheels/cpe-1.2.1-py3-none-any.whl
#17 11.12 Processing /tmp/wheels/packageurl_python-0.9.4-py2.py3-none-any.whl
#17 11.14 Processing /tmp/wheels/django_crum-0.7.9-py2.py3-none-any.whl
#17 11.16 Processing /tmp/wheels/JSON_log_formatter-0.3.0-py3-none-any.whl
#17 11.18 Processing /tmp/wheels/django_split_settings-1.0.1-py3-none-any.whl
#17 11.20 Processing /tmp/wheels/django_debug_toolbar-3.2-py3-none-any.whl
#17 11.22 Processing /tmp/wheels/django_debug_toolbar_request_history-0.1.3-py2.py3-none-any.whl
#17 11.24 Processing /tmp/wheels/vcrpy-4.1.1-py2.py3-none-any.whl
#17 11.26 Processing /tmp/wheels/vcrpy_unittest-0.1.7-py2.py3-none-any.whl
#17 11.27 Processing /tmp/wheels/django_tagulous-1.1.0-py3-none-any.whl
#17 11.29 Processing /tmp/wheels/PyJWT-2.0.1-py3-none-any.whl
#17 11.31 Processing /tmp/wheels/cvss-2.2-py2.py3-none-any.whl
#17 11.33 Processing /tmp/wheels/django_jsonfield_backport-1.0.3-py3-none-any.whl
#17 11.35 Processing /tmp/wheels/django_fieldsignals-0.6.0-py2.py3-none-any.whl
#17 11.37 Processing /tmp/wheels/hyperlink-21.0.0-py2.py3-none-any.whl
#17 11.40 Processing /tmp/wheels/packaging-20.9-py2.py3-none-any.whl
#17 11.42 Processing /tmp/wheels/six-1.15.0-py2.py3-none-any.whl
#17 11.44 Processing /tmp/wheels/webencodings-0.5.1-py2.py3-none-any.whl
#17 11.53 Processing /tmp/wheels/billiard-3.6.4.0-py3-none-any.whl
#17 11.55 Processing /tmp/wheels/vine-1.3.0-py2.py3-none-any.whl
#17 11.57 Processing /tmp/wheels/kombu-4.6.11-py2.py3-none-any.whl
#17 11.63 Processing /tmp/wheels/cffi-1.14.5-cp36-cp36m-manylinux1_x86_64.whl
#17 11.67 Processing /tmp/wheels/sqlparse-0.4.1-py3-none-any.whl
#17 11.70 Processing /tmp/wheels/django_jsonfield-1.4.1-py2.py3-none-any.whl
#17 11.81 Processing /tmp/wheels/django_appconf-1.0.4-py2.py3-none-any.whl
#17 11.83 Processing /tmp/wheels/pilkit-2.0-py3-none-any.whl
#17 11.88 Processing /tmp/wheels/prometheus_client-0.10.1-py2.py3-none-any.whl
#17 11.90 Processing /tmp/wheels/simplejson-3.17.2-cp36-cp36m-manylinux2010_x86_64.whl
#17 11.92 Processing /tmp/wheels/coreapi-2.3.3-py2.py3-none-any.whl
#17 11.94 Processing /tmp/wheels/openapi_codec-1.3.2-py3-none-any.whl
#17 11.97 Processing /tmp/wheels/djangorestframework_jwt-1.11.0-py2.py3-none-any.whl
#17 11.99 Processing /tmp/wheels/django_rest_auth-0.9.5-py2.py3-none-any.whl
#17 12.01 Processing /tmp/wheels/pysaml2-6.5.1-py2.py3-none-any.whl
#17 12.07 Processing /tmp/wheels/python_mimeparse-1.6.0-py2.py3-none-any.whl
#17 12.12 Processing /tmp/wheels/ruamel.yaml-0.17.4-py3-none-any.whl
#17 12.14 Processing /tmp/wheels/inflection-0.5.1-py2.py3-none-any.whl
#17 12.16 Processing /tmp/wheels/uritemplate-3.0.1-py2.py3-none-any.whl
#17 12.19 Processing /tmp/wheels/coreschema-0.0.4-py3-none-any.whl
#17 12.22 Processing /tmp/wheels/gitdb-4.0.7-py3-none-any.whl
#17 12.25 Processing /tmp/wheels/google_auth_httplib2-0.1.0-py2.py3-none-any.whl
#17 12.27 Processing /tmp/wheels/google_api_core-1.26.3-py2.py3-none-any.whl
#17 12.30 Processing /tmp/wheels/httplib2-0.19.1-py3-none-any.whl
#17 12.34 Processing /tmp/wheels/rsa-4.7.2-py3-none-any.whl
#17 12.36 Processing /tmp/wheels/pyasn1_modules-0.2.8-py2.py3-none-any.whl
#17 12.38 Processing /tmp/wheels/cachetools-4.2.1-py3-none-any.whl
#17 12.38 Requirement already satisfied: setuptools>=40.3.0 in /usr/local/lib/python3.6/site-packages (from google-auth==1.28.0->-r ./requirements.txt (line 57)) (50.3.0)
#17 12.41 Processing /tmp/wheels/requests_oauthlib-1.3.0-py2.py3-none-any.whl
#17 12.49 Processing /tmp/wheels/idna-2.10-py2.py3-none-any.whl
#17 12.55 Processing /tmp/wheels/oauthlib-3.1.0-py2.py3-none-any.whl
#17 12.57 Processing /tmp/wheels/requests_toolbelt-0.9.1-py2.py3-none-any.whl
#17 12.59 Processing /tmp/wheels/pbr-5.5.1-py2.py3-none-any.whl
#17 12.65 Processing /tmp/wheels/importlib_metadata-3.10.0-py3-none-any.whl
#17 12.69 Processing /tmp/wheels/et_xmlfile-1.0.1-py3-none-any.whl
#17 12.77 INFO: pip is looking at multiple versions of ptvsd to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of psycopg2-binary to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of pillow to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of pdfkit to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of packageurl-python to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of openpyxl to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of mysqlclient to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of markdown to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of lxml to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of json-log-formatter to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of jira to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of hyperlink to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of humanize to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of html2text to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of gunicorn to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of google-auth-oauthlib to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of google-auth to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of google-api-python-client to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of gitpython to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of drf-yasg to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of djangorestframework to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of django-watson to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of django-tastypie to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of django-tagulous to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of django-tagging to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of django-split-settings to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of django-slack to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of django-saml2-auth to determine which version is compatible with other requirements. This could take a while.
#17 12.77 INFO: pip is looking at multiple versions of django-rest-swagger to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-prometheus to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-polymorphic to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-multiselectfield to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-jsonfield-backport to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-imagekit to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-filter to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-fieldsignals to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-extensions to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-environ to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-debug-toolbar-request-history to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-debug-toolbar to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-dbbackup to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-crum to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-crispy-forms to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-celery-results to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-auditlog to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of defusedxml to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of cvss to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of cryptography to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of cpe to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of coverage to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of celery to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of bleach to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of <Python from Requires-Python> to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of asteval to determine which version is compatible with other requirements. This could take a while.
#17 12.78 INFO: pip is looking at multiple versions of django-tastypie-swagger to determine which version is compatible with other requirements. This could take a while.
#17 12.78 ERROR: Cannot install -r ./requirements.txt (line 30) and PyJWT==2.0.1 because these package versions have conflicting dependencies.
#17 12.78
#17 12.78 The conflict is caused by:
#17 12.78     The user requested PyJWT==2.0.1
#17 12.78     pygithub 1.54.1 depends on pyjwt<2.0
#17 12.78
#17 12.78 To fix this you could try to:
#17 12.78 1. loosen the range of package versions you've specified
#17 12.78 2. remove package versions to allow pip attempt to solve the dependency conflict
#17 12.78
#17 12.78 ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/user_guide/#fixing-conflicting-dependencies
```

**Comment by damiencarol** [comment]
@dvelardez you are trying to build images for `1.14.1` manually, right? Why don't you use images from Docker Hub?

**Comment by dvelardez** [comment]
Exactly.
Thanks in advance.

**Comment by dvelardez** [comment]
Sorry, I just reread the email. I am trying to compile locally to upload the image to a private repository.
Thanks!El 13 abr. 2021 07:18, Damien Carol ***@***.***> escribió:
@dvelardez you are trying to build images for 1.14.1 manually, right? Why don't you use images from Docker Hub?

**Comment by valentijnscholten** [comment]
The question is why are builds succeeding for us and in GitHub Actions, but not for people locally? The python base image we are using is or should still use the old pip dependency resolver which still works with our requirements.txt contents.
We are unable to support the new resolver as there are upstream packages causing problems without any fix being available yet.

**Comment by Brain2998** [comment]
>  should still use the old pip dependency resolver which still works with our requirements.txt contents

Could you please specify version of pip with correct resolver?

**Comment by damiencarol** [comment]
I think we don't use the best path here.
I mean if you really want to control the loading of the images and use a specific registry, you can force docker command to do that for you.

```
[damien@damien ~]$ docker pull defectdojo/defectdojo-django:1.14.1
1.14.1: Pulling from defectdojo/defectdojo-django
Digest: sha256:89896f878ac557a56ca28440ec26e657bc995a3dee22f3ce8959cb9f1c0a6c2e
Status: Image is up to date for defectdojo/defectdojo-django:1.14.1
docker.io/defectdojo/defectdojo-django:1.14.1
[damien@damien ~]$ docker save defectdojo/defectdojo-django:1.14.1 -o defectdojo-django_1.14.1.tar
[damien@damien ~]$ file defectdojo-django_1.14.1.tar
defectdojo-django_1.14.1.tar: POSIX tar archive
```
This way you can keep it somewhere, even check the digest and bu sure that it's the official image.

You load it to your registry this way:
```
docker login xxx xxx xxx
docker load -i defectdojo-django_1.14.1.tar
```

**Comment by valentijnscholten** [comment]
There might be people making modifications and wanting to build their own images. The question is why is he getting this error and we don't. @dvelardez can you detail all the changes made to the source before trying to build them?

In the mean time the PyGitHub package finally released an update addressing this, so chances are in the next release we will be compatible with the new pip dependency resolver.

In the mean time you can use `pip install -r requirements.txt --use-deprecated legacy-resolver`

**Comment by dvelardez** [comment]
I have not made any changes to the source code.
Thanks for the workaround.

**Comment by valentijnscholten** [comment]
Something must be different because the stock sourcecode and dockerfile are building correctly.

**Comment by valentijnscholten** [comment]
This is now resolved in dev / 2.0.0.
